### PR TITLE
feat: support websocket reverse proxy

### PR DESCRIPTION
**Pull request by justlorain**
#### What type of PR is this?

feat

#### What this PR does / why we need it (English/Chinese):

Support reverse proxy for websocket

#### Which issue(s) this PR fixes:

Fixes (https://github.com/cloudwego/hertz/issues/926)

### CODE REVIEW

**Comment by li-jin-gou** [comment]
use `gofumpt -e -d -w -extra .`

File: go.mod [hunk 1]
```diff
@@ -2,4 +2,8 @@ module github.com/hertz-contrib/reverseproxy
 
 go 1.16
 
-require github.com/cloudwego/hertz v0.6.5
+require (
+	github.com/cloudwego/hertz v0.6.5
+	github.com/gorilla/websocket v1.5.1
```
Comment (li-jin-gou) [inline go.mod:8]: add license

File: README.md [hunk 2]
```diff
@@ -86,5 +86,28 @@ func main() {
 
 `ReverseProxy` provides `SetDirector`、`SetModifyResponse`、`SetErrorHandler` to modify `Request` and `Response`.
 
+### Websocket Reverse Proxy
```
Comment (li-jin-gou) [inline README.md:95]: 如果是参考 https://github.com/yeqown/fasthttp-reverse-proxy/blob/master/ws_reverseproxy.go 的话注明一下出处并且添加一下 license

Comment (justlorain) [reply README.md:95]: done, thx

File: go.mod (hunk 1 above)
Comment (justlorain) [reply go.mod:8]: done, thx

File: ws_reverse_proxy.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+	header := make(http.Header)
+	hzHeader.VisitAll(func(key, value []byte) {
+		k, v := string(key), string(value)
+		header.Add(k, v)
+	})
+	return header
+}
+
+func prepareForwardHeader(_ context.Context, c *app.RequestContext) *protocol.RequestHeader {
+	forwardHeader := &protocol.RequestHeader{}
```
Comment (li-jin-gou) [inline ws_reverse_proxy.go]: 感觉这没必要用 hertz 的 header

Comment (li-jin-gou) [reply ws_reverse_proxy.go]: 这次拷贝到 http.Header 是迟早的事情

File: ws_reverse_proxy.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+			case err = <-errClientC:
+				errMsg = "copy websocket response err: %v"
+			case err = <-errBackendC:
+				errMsg = "copy websocket request err: %v"
+			}
+
+			var ce *websocket.CloseError
+			var hzce *hzws.CloseError
+			if !errors.As(err, &ce) || !errors.As(err, &hzce) {
+				hlog.Errorf(errMsg, err)
```
Comment (li-jin-gou) [inline ws_reverse_proxy.go]: 统一 hlog.Ctx***  保证链路追踪可以串起来

**Review by li-jin-gou — APPROVED** [review]

**Pull request merged**